test/functional/test_framework/util.py

```diff
@@ -307,26 +308,63 @@ def initialize_chain_clean(test_dir, num_nodes):
     for i in range(num_nodes):
         datadir=initialize_datadir(test_dir, i)
 
+def start_nodes(num_nodes, dirname, extra_args=None, rpchost=None, timewait=None, binary=None):
+    """Start multiple bitcoinds and return RPC connections to them.
+
+    This function uses an asyncio event loop to start the bitcoind nodes in parallel to speed
+    up test execution."""
+    if extra_args is None: extra_args = [ None for _ in range(num_nodes) ]
+    if binary is None: binary = [ None for _ in range(num_nodes) ]
+    rpcs = []
+    if num_nodes > 0:
+        try:
+            loop = asyncio.get_event_loop()
+            tasks = [asyncio.ensure_future(start_node_async(i, dirname, extra_args[i], rpchost, timewait=timewait, binary=binary[i])) for i in range(num_nodes)]
+            loop.run_until_complete(asyncio.wait(tasks))
+            rpcs = [task.result() for task in tasks]
+        except: # If one node failed to start, stop the others
+            stop_nodes(rpcs)
+            raise
+    return rpcs
 
 def start_node(i, dirname, extra_args=None, rpchost=None, timewait=None, binary=None, stderr=None):
-    """
-    Start a bitcoind and return RPC connection to it
-    """
+    """Start a single bitcoind node and return an RPC connection to it."""
+    loop = asyncio.get_event_loop()
+    task = asyncio.ensure_future(start_node_async(i, dirname, extra_args, rpchost, timewait, binary, stderr))
+    loop.run_until_complete(task)
+    return task.result()
+
+async def start_node_async(i, dirname, extra_args=None, rpchost=None, timewait=None, binary=None, stderr=None):
+    """An asynchronous co-routine that starts a bitcoind node and returns an RPC connection to it."""
     datadir = os.path.join(dirname, "node"+str(i))
     if binary is None:
         binary = os.getenv("BITCOIND", "bitcoind")
     args = [ binary, "-datadir="+datadir, "-server", "-keypool=1", "-discover=0", "-rest", "-logtimemicros", "-debug", "-mocktime="+str(get_mocktime()) ]
     if extra_args is not None: args.extend(extra_args)
     bitcoind_processes[i] = subprocess.Popen(args, stderr=stderr)
-    logger.debug("initialize_chain: bitcoind started, waiting for RPC to come up")
+    logger.debug("initialize_chain: bitcoind %d started, waiting for RPC to come up" % i)
     url = rpc_url(i, rpchost)
-    wait_for_bitcoind_start(bitcoind_processes[i], url, i)
-    logger.debug("initialize_chain: RPC successfully started")
+
+    while True:
+        if bitcoind_processes[i].poll() is not None:
```

Comment:
This whole block of code except the sleep is duplicated from wait_for_bitcoind_start. It would be good to factor out the common code into a function to avoid duplication.